foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java

```diff
@@ -0,0 +1,318 @@
+/**
+ * Copyright (c) 2012 - 2016 YCSB contributors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package com.yahoo.ycsb.db;
+
+import com.apple.foundationdb.*;
+import com.apple.foundationdb.async.AsyncIterable;
+import com.apple.foundationdb.tuple.Tuple;
+
+import com.yahoo.ycsb.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.helpers.MessageFormatter;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * FoundationDB client for YCSB framework.
+ */
+
+public class FoundationDBClient extends DB {
+  private FDB fdb;
+  private Database db;
+  private String dbName;
+  private int batchSize;
+  private int batchCount;
+  private static final String API_VERSION          = "foundationdb.apiversion";
+  private static final String API_VERSION_DEFAULT  = "520";
+  private static final String CLUSTER_FILE         = "foundationdb.clusterfile";
+  private static final String CLUSTER_FILE_DEFAULT = "./fdb.cluster";
+  private static final String DB_NAME              = "foundationdb.dbname";
+  private static final String DB_NAME_DEFAULT      = "DB";
+  private static final String DB_BATCH_SIZE_DEFAULT = "0";
+  private static final String DB_BATCH_SIZE         = "foundationdb.batchsize";
+
+  private Vector<String> batchKeys;
+  private Vector<Map<String, ByteIterator>> batchValues;
+
+  private static Logger logger = LoggerFactory.getLogger(FoundationDBClient.class);
+
+  private static final AtomicInteger INIT_COUNT = new AtomicInteger(0);
+
+  /**
+   * Initialize any state for this DB. Called once per DB instance; there is one DB instance per client thread.
+   */
+  @Override
+  public void init() throws DBException {
+    INIT_COUNT.incrementAndGet();
+    synchronized (INIT_COUNT) {
+      // initialize FoundationDB driver
+      final Properties props = getProperties();
+      String apiVersion = props.getProperty(API_VERSION, API_VERSION_DEFAULT);
+      String clusterFile = props.getProperty(CLUSTER_FILE, CLUSTER_FILE_DEFAULT);
+      String dbBatchSize = props.getProperty(DB_BATCH_SIZE, DB_BATCH_SIZE_DEFAULT);
+      dbName = props.getProperty(DB_NAME, DB_NAME_DEFAULT);
+
+      logger.info("API Version: {}", apiVersion);
+      logger.info("Cluster File: {}\n", clusterFile);
+
+      try {
+        fdb = FDB.selectAPIVersion(Integer.parseInt(apiVersion.trim()));
```

Comment:
is this the bit that needs to be protected by the synchronized block? is the db object itself threadsafe?

Reply:
Thank you for the comment. The synchronized block is not needed. I've removed it from the PR.